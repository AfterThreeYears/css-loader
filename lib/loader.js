/*
  MIT License http://www.opensource.org/licenses/mit-license.php
  Author Tobias Koppers @sokra
*/
const loaderUtils = require("loader-utils");
const postcss = require("postcss");
const plugin = require("./plugin");
const getImportPrefix = require("./getImportPrefix");
const SyntaxError = require("./SyntaxError");

module.exports = function(content, map) {
  const options = loaderUtils.getOptions(this) || {};

  // Todo validate options

  const cb = this.async();
  const sourceMap = options.sourceMap;

  if (sourceMap && map) {
    if (typeof map === "string") {
      map = JSON.parse(map);
    }

    if (map.sources) {
      map.sources = map.sources.map(source => source.replace(/\\/g, "/"));
      map.sourceRoot = "";
    }
  } else {
    // Some loaders (example `"postcss-loader": "1.x.x"`) always generates source map, we should remove it
    map = null;
  }

  // We need a prefix to avoid path rewriting of PostCSS
  const from =
    "/css-loader!" +
    loaderUtils
      .getRemainingRequest(this)
      .split("!")
      .pop();
  const to = loaderUtils
    .getCurrentRequest(this)
    .split("!")
    .pop();

  postcss([
    plugin({
      url: options.url !== false,
      import: options.import !== false
    })
  ])
    .process(content, {
      from,
      to,
      map: sourceMap
        ? {
            prev: map,
            sourcesContent: true,
            inline: false,
            annotation: false
          }
        : null
    })
    .then(result => {
      var cssAsString = JSON.stringify(result.css);

      if (options.import !== false) {
        var alreadyImported = {};
        var importJs = result.messages
          .filter(message => message.type === "at-rule-import")
          .filter(imp => {
            if (!imp.mediaQuery) {
              if (alreadyImported[imp.url]) {
                return false;
              }

              alreadyImported[imp.url] = true;
            }

            return true;
          })
          .map(imp => {
            if (!loaderUtils.isUrlRequest(imp.url)) {
              return (
                "exports.push([module.id, " +
                JSON.stringify("@import url(" + imp.url + ");") +
                ", " +
                JSON.stringify(imp.mediaQuery) +
                "]);"
              );
            }

            // for importing CSS
            var importUrlPrefix = getImportPrefix(this, options);
            var importUrl = importUrlPrefix + imp.url;

            return (
              "exports.i(require(" +
              loaderUtils.stringifyRequest(this, importUrl) +
              "), " +
              JSON.stringify(imp.mediaQuery) +
              ");"
            );
          })
          .join("\n");
      }

      // Helper for ensuring valid CSS strings from requires
      let urlEscapeHelper = "";

      if (options.url !== false) {
        urlEscapeHelper =
          "var runtimeEscape = require(" +
          loaderUtils.stringifyRequest(
            this,
            require.resolve("./runtimeEscape.js")
          ) +
          ");\n";

        result.messages
          .filter(message => message.type === "css-loader-import-url")
          .forEach(message => {
            const { placeholder, url } = message;
            const splittedURL = url.split(/(\?)?#/);
            const importURLString =
              '" + runtimeEscape(require(' +
              loaderUtils.stringifyRequest(this, splittedURL[0]) +
              ')) + "' +
              (splittedURL[1] ? splittedURL[1] : "") +
              (splittedURL[2] ? `#${splittedURL[2]}` : "");

            cssAsString = cssAsString.replace(placeholder, importURLString);
          });
      }

      // Todo need save backward compatibility with old `style-loader`
      var exportJs = "";

      if (exportJs) {
        exportJs = "exports.locals = " + exportJs + ";";
      }

      var moduleJs;

      if (sourceMap && result.map) {
        map = result.map.toJSON();

        if (map.sources) {
          map.sources = map.sources.map(source =>
            source
              .split("!")
              .pop()
              .replace(/\\/g, "/")
          );
          map.sourceRoot = "";
        }

        map.file = map.file
          .split("!")
          .pop()
          .replace(/\\/g, "/");
        map = JSON.stringify(map);

        moduleJs =
          "exports.push([module.id, " + cssAsString + ', "", ' + map + "]);";
      } else {
        moduleJs = "exports.push([module.id, " + cssAsString + ', ""]);';
      }

      // embed runtime
      cb(
        null,
        urlEscapeHelper +
          "exports = module.exports = require(" +
          loaderUtils.stringifyRequest(this, require.resolve("./runtime.js")) +
          ")(" +
          sourceMap +
          ");\n" +
          "// imports\n" +
          importJs +
          "\n\n" +
          "// module\n" +
          moduleJs +
          "\n\n" +
          "// exports\n" +
          exportJs
      );
    })
    .catch(err => {
      // Todo if (err.file) this.addDependency(err.file)

      cb(err.name === "CssSyntaxError" ? new SyntaxError(err) : err);
    });
};
