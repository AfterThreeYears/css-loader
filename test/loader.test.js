/* global describe it expect */

import fs from 'fs';
import loader from '../src/';
import path from 'path';

describe('loader', () => {
  const fixtures = path.resolve(__dirname, 'loader-fixtures');
  const tests = fs.readdirSync(fixtures).filter((name) => /\.css$/.test(name));
  tests.forEach((filename) => {
    it(`should process ${filename}`, () => {
      const filepath = path.resolve(fixtures, filename);
      const content = fs.readFileSync(filepath, 'utf-8').replace(/\r\n?/g, '\n');
      let result;
      const directResult = loader.call({
        request: "css-loader!file.css",
        loaders: [
          {
            path: "css-loader",
          },
        ],
        context: __dirname,
        currentLoaderIndex: 0,
        sourceMap: false,
        callback(err, ...args) {
          if(err) {
            throw err;
          }
          result = args;
        }
      }, content);
      result = result || [directResult];
      expect(result).toMatchSnapshot();
    });
  });
});
