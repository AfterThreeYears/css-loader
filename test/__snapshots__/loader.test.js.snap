// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader should process import.css 1`] = `
Array [
  "// css runtime
import * as runtime from \\"../runtime/index.js\\";

// declarations
import { ABC as cssLoaderImport0_ABC } from \\"some-file\\";
import { DEF as cssLoaderImport1_DEF } from \\"some-file\\";
import { HIJ as cssLoaderImport2_HIJ } from \\"other-file\\";
import cssLoaderImport3 from \\"some-file\\";
import cssLoaderImport4 from \\"other-file\\";

// CSS
export default runtime.create([
  cssLoaderImport3,
  cssLoaderImport4,
  runtime.moduleWithoutSourceMap(module.id, \\"body {\\\\n  background: \\" + cssLoaderImport0_ABC + \\" \\" + cssLoaderImport1_DEF + \\" \\" + cssLoaderImport2_HIJ + \\" klm\\\\n}\\\\n\\\\n.\\" + cssLoaderImport0_ABC + \\" { \\" + cssLoaderImport1_DEF + \\": \\" + cssLoaderImport2_HIJ + \\"; }\\\\n\\")
]);",
]
`;

exports[`loader should process simple.css 1`] = `
Array [
  "// css runtime
import * as runtime from \\"../runtime/index.js\\";

// declarations


// CSS
export default runtime.create([
  runtime.moduleWithoutSourceMap(module.id, \\"body {\\\\n  background: red;\\\\n}\\\\n\\\\n.class {\\\\n  color: rgb(1, 2, 3);\\\\n}\\\\n\\")
]);",
]
`;
